src/mbgl/style/conversion/filter.cpp

```diff
@@ -1,11 +1,48 @@
 #include <mbgl/style/conversion/filter.hpp>
 #include <mbgl/util/geometry.hpp>
+#include <mbgl/style/expression/expression.hpp>
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/style/conversion/expression.hpp>
 
 namespace mbgl {
 namespace style {
 namespace conversion {
+    
+using GeometryValue = mapbox::geometry::value;
 
-static optional<Value> normalizeValue(const optional<Value>& value, Error& error) {
+// This is a port from https://github.com/mapbox/mapbox-gl-js/blob/master/src/style-spec/feature_filter/index.js
+static bool isExpressionFilter(const Convertible& filter) {
+    if (!isArray(filter) || arrayLength(filter) == 0) {
+        return false;
+    }
+    
+    optional<std::string> op = toString(arrayMember(filter, 0));
+    
+    if (*op == "has") {
+        auto operand = toString(arrayMember(filter, 1));
+        return arrayLength(filter) >= 2 && operand && *operand != "$id" && *operand != "$type";
```

Comment:
The `arrayLength` check needs to come before `arrayMember`. Also needs a guard check for string type.

Reply:
:+1:

```c++
optional<std::string> operand = toString(arrayMember(filter, 1));
return operand && *operand != "$id" && *operand != "$type";
```

My understanding is that checking the truthiness of `operand` serves as a guard check for string type. Is that the case?

Reply:
Oh, you're right, I missed that condition.